### DISCUSSION TITLE
Which scenarios can cause this assertion to be incorrect?

### DISCUSSION THREAD

**Discussion by algebraic-dev** [libuv/libuv#4628] [General]
```
Assertion failed: (!(handle->flags & UV_HANDLE_CLOSED)), function uv__finish_close, file core.c, line 313.
```

I'm having an issue when using an event loop thread, an async waker, and multiple mutexes to manage access to the event loop, similar to the approach used in Julia so multiple threads can register events in the event loop. The error occurs when calling uv_close, which typically happens inside a timer callback. It seems like uv_close is being executed while holding the event loop mutex, inside the event loop thread potentially within the uv_run function. Could this be causing a deadlock or some other issue? Why is this error happening, and how can I resolve it?

**Comment by saghul** [comment]
Looks like a potential double-close.

Any chance you are calling uv_close from 2 separate threads?

**Reply by bnoordhuis** [reply]
Rephrased: call uv_close() from the same thread that uv_run() runs on.

**Reply by algebraic-dev** [reply]
> Looks like a potential double-close.
> 
> Any chance you are calling uv_close from 2 separate threads?

I changed the code to just run the uv_close when it calls the callback function, so it probably is always running in the event loop thread, I'm creating a lot of timers and then trying to use uv_close fast.

So idk, I'm not so sure if it's a double close.

In lldb this is the only thing i get

```
* thread #2, stop reason = hit program assert
    frame #0: 0x00000001916255f0 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000019165dc20 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x000000019156aa30 libsystem_c.dylib`abort + 180
    frame #3: 0x0000000191569d20 libsystem_c.dylib`__assert_rtn + 284
  * frame #4: 0x0000000100b5eea4 libuv.1.dylib`uv_run.cold.1 + 40
    frame #5: 0x0000000100b4e2c0 libuv.1.dylib`uv_run + 672
    frame #6: 0x00000001004c0f80 ata`std::__1::__function::__func<initialize_libuv::$_1, std::__1::allocator<initialize_libuv::$_1>, void ()>::operator()() + 96
    frame #7: 0x0000000100495eec ata`lean::lthread::imp::_main(void*) + 56
    frame #8: 0x000000019165df94 libsystem_pthread.dylib`_pthread_start + 136
```
